llvm/lib/SYCLLowerIR/ModuleSplitter.cpp

```diff
@@ -660,11 +660,36 @@ void ModuleDesc::restoreLinkageOfDirectInvokeSimdTargets() {
   }
 }
 
-// Predicate for Internalize pass.
-bool mustPreserveGV(const GlobalValue &GV) {
-  if (const Function *F = dyn_cast<Function>(&GV))
-    if (!canBeImportedFunction(*F))
-      return false;
+// Predicate for Internalize pass. The pass is very aggressive and essentially
+// tries to internalize absolutely everything. This function serves as "input
+// from a linker" that tells the pass what must be preserved in order to make
+// the transformation safe.
+static bool mustPreserveGV(const GlobalValue &GV) {
+  if (const Function *F = dyn_cast<Function>(&GV)) {
+    // When dynamic linking is supported, we internalize everything that can
+    // not be imported which also means that there is no point of having it
+    // visible outside of the current module.
+    if (SupportDynamicLinking)
+      return canBeImportedFunction(*F);
+
+    // Otherwise, we are being even more aggressive: SYCL modules are expected
+    // to be self-contained, meaning that they have no external dependencies.
+    // Therefore, we can internalize every function that is not an entry point.
+    // One exception here is virtual functions, because their implementation
+    // relies on dynamic linking and therefore they must be preserved if present
```

Comment:
Is "dynamic linking" in line 679 the same as "dynamic linking" in 669?  

Some readers of these comments may expect "dynamic linking" to always take the return path on line 673 and should not be relevant at line 674.

Reply:
> Is "dynamic linking" in line 679 the same as "dynamic linking" in 669?

Both yes and no :) It is the same in context that SYCL RT will collect several device images and link them together at runtime. It is not the same, because it is done using a different algorithm for searching for device images that should be linked together.

I will try to clarify the comment to better clarify the distinction between the two.

> Since the code at 685 is necessary for virtual functions, does that mean that virtual functions cannot work if SupportDynamicLinking is on (i.e. sycl-post-link was invoked with -support-dynamic-linking (i.e. the user used -fallow-device-dependencies))?

Those two features should be able to co-exist, even though we will need to write more code if we want a virtual function defined in a shared library to be callable from an application - that won't work right now. As long as virtual functions do not cross application/shared library boundary, they should work even if shared libraries are involved.

Note that a virtual function that is callable from device will meet condition of an "exported" function and it won't be internalized by `SupportDynamicLinking` code path. It is already tested by:
https://github.com/intel/llvm/blob/d8939948ca62955613e51d63e7c08d683c18d529/llvm/test/tools/sycl-post-link/virtual-functions/module-cleanup.ll#L6

But I will add some tests to check co-existence of virtual functions and shared libraries later during virtual functions test suite expansion.

Reply:
> > Is "dynamic linking" in line 679 the same as "dynamic linking" in 669?
> 
> Both yes and no :) It is the same in context that SYCL RT will collect several device images and link them together at runtime. It is not the same, because it is done using a different algorithm for searching for device images that should be linked together.
> 
> I will try to clarify the comment to better clarify the distinction between the two.

I've pushed 95d2649b9897f73ae7d090ef59266721e6c48a47 with the comment updated, could you please take a look and see if its clearer now?

Reply:
Comments look good to me.